compiler/src/main.rs

```diff
@@ -258,45 +262,31 @@ fn run(options: CandyRunOptions) {
     };
 
     info!("Running main function.");
-    // TODO: Add environment stuff.
+    // TODO: Add more environment stuff.
     let mut vm = Vm::new();
-    let stdout = vm.create_channel();
+    let mut stdout = StdoutService::new(&mut vm);
     let environment = {
         let stdout_symbol = heap.create_symbol("Stdout".to_string());
-        let stdout_port = heap.create_send_port(stdout);
-        heap.create_struct([(stdout_symbol, stdout_port)].into_iter().collect())
+        let stdout_port = heap.create_send_port(stdout.channel);
+        heap.create_struct(HashMap::from([(stdout_symbol, stdout_port)]))
     };
     vm.set_up_for_running_closure(heap, main, &[environment]);
     loop {
         info!("Tree: {:#?}", vm);
         match vm.status() {
             Status::CanRun => {
-                debug!("VM still running.");
+                debug!("VM is still running.");
                 vm.run(&mut DbUseProvider { db: &db }, &mut RunForever);
-                // TODO: handle operations
             }
             Status::WaitingForOperations => {
                 todo!("VM can't proceed until some operations complete.");
             }
             _ => break,
         }
-        let stdout_operations = vm
-            .external_operations
-            .get_mut(&stdout)
-            .unwrap()
-            .drain(..)
-            .collect_vec();
-        for operation in stdout_operations {
-            match operation {
-                vm::Operation::Send {
-                    performing_fiber,
-                    packet,
-                } => {
-                    info!("Sent to stdout: {}", packet.value.format(&packet.heap));
-                    vm.complete_send(performing_fiber);
-                }
-                vm::Operation::Receive { .. } => unreachable!(),
-                vm::Operation::Drop => vm.free_channel(stdout),
+        stdout.run(&mut vm);
+        for channel in vm.unreferenced_channels.iter().copied().collect_vec() {
+            if channel != stdout.channel {
+                vm.free_channel(channel);
             }
```

Comment:
`stdout` should behave like other channels and once it's no longer referenced, it won't be passed to the VM again.

Also, I'm not sure whether `.collect_vec()` is needed for the `for` loop's iterable

Reply:
Fair enough. The `collect_vec` is actually needed though because the `free_channel` method needs a mutable VM. It also removes the channel from the `unreferenced_channels` (which would not be a good idea to do while we are iterating over that).